File: backend/src/main/kotlin/com/clubs/eventtemplate/EventTemplate.kt
package com.clubs.eventtemplate

import com.clubs.event.EventFormat
import java.time.LocalTime
import java.time.OffsetDateTime
import java.util.UUID

/**
 * Именованная заготовка формы создания встречи, принадлежащая клубу (V79).
 *
 * Хранит всё, что организатор заполняет при создании события, КРОМЕ даты — она меняется
 * всегда. Вместо даты живут [defaultWeekday] и [defaultTime]: форма подставляет ближайшее
 * будущее совпадение, но поле остаётся обычным, редактируемым.
 *
 * Шаблон не является ни расписанием (авто-создание встреч сюда не входит), ни новым форматом
 * события: применение открывает ту же форму и тот же `POST /api/clubs/{id}/events`.
 * Спека: docs/modules/event-templates.md.
 */
data class EventTemplate(
    val id: UUID,
    val clubId: UUID,
    // Ярлык списка выбора, отдельный от title: у двух шаблонов может совпадать название встречи.
    val name: String,
    val title: String,
    val description: String?,
    val locationText: String?,
    // Точка места (WGS-84). Инвариант пары зеркалит события: оба заданы или оба null.
    val locationLat: Double?,
    val locationLon: Double?,
    val locationHint: String?,
    // Потолок мест; null = шаблон открытой встречи.
    val participantLimit: Int?,
    // Минимум участников (V86); шаблон ЗАПОМИНАЕТ включённый минимум и подставляет его в форму —
    // так клуб, который живёт с кворумами, получает его по умолчанию. null = выключен.
    val minParticipants: Int?,
    val format: EventFormat,
    // Свой интервал Этапа 2 в минутах; null = глобальный дефолт бэкенда.
    val stage2LeadMinutes: Int?,
    val photoUrl: String?,
    // День недели повторов (1 = понедельник … 7 = воскресенье, ISO-8601) и время — в ЛОКАЛЬНОЙ
    // зоне организатора. Их считает клиент: event_datetime хранится как TIMESTAMPTZ, и вывод
    // дня недели из UTC сдвинул бы «вторник 19:00» в другой день. null = не задано.
    val defaultWeekday: Short?,
    val defaultTime: LocalTime?,
    val createdBy: UUID,
    val createdAt: OffsetDateTime?,
    val updatedAt: OffsetDateTime?
)

/** Шаблон вместе с именем клуба-владельца — кросс-клубовый список в пикере «+» показывает клуб. */
data class EventTemplateWithClub(
    val template: EventTemplate,
    val clubName: String
)

/**
 * Содержимое шаблона БЕЗ имени-ярлыка — то, по чему два шаблона считаются одинаковыми
 * (требование PO 2026-08-11: «Встреча в парке» и «Сходить в парк» с идентичными полями —
 * это один шаблон под двумя именами, и в списке они неразличимы).
 *
 * Отдельный тип, а не сравнение полей вручную: `data class` даёт `equals` даром, и добавленное
 * поле шаблона нельзя молча забыть в сравнении — компилятор потребует заполнить его в обоих
 * конструкторах ([EventTemplateMapper.toContent]).
 */
data class EventTemplateContent(
    val title: String,
    val description: String?,
    val locationText: String?,
    val locationLat: Double?,
    val locationLon: Double?,
    val locationHint: String?,
    val participantLimit: Int?,
    val minParticipants: Int?,
    val format: EventFormat,
    val stage2LeadMinutes: Int?,
    val photoUrl: String?,
    val defaultWeekday: Short?,
    val defaultTime: LocalTime?
)

File: frontend/src/utils/eventTemplate.ts
import { toDatetimeLocalValue } from './formatters';
import type { EventTemplateDto, SaveEventTemplateBody } from '../api/eventTemplates';
import type { EventFormat } from '../types/api';

/**
 * Помощники шаблонов встреч: вывод «дня недели + времени» из конкретной встречи и обратная
 * подстановка ближайшей будущей даты.
 *
 * ВСЁ считается в ЛОКАЛЬНОЙ зоне устройства — и это принципиально. `event_datetime` хранится
 * как TIMESTAMPTZ, организатор выбирает время в `datetime-local`, то есть по своим настенным
 * часам. Выводи мы день недели из UTC — «вторник 19:00» в Москве стал бы «вторником 16:00»,
 * и подстановка промахивалась бы днём при вечерних встречах.
 * Спека: docs/modules/event-templates.md § 4.3.
 */

/** Понедельник = 1 … воскресенье = 7 (ISO-8601). JS `getDay()` считает с воскресенья-0. */
export function isoWeekdayOf(date: Date): number {
  return ((date.getDay() + 6) % 7) + 1;
}

/** Локальное «HH:mm:ss» — формат, который принимает и отдаёт бэкенд (java.time.LocalTime). */
export function localTimeOf(date: Date): string {
  const pad = (n: number) => String(n).padStart(2, '0');
  return `${pad(date.getHours())}:${pad(date.getMinutes())}:00`;
}

/**
 * Ближайший будущий момент, попадающий на [weekday] в [time], как значение для
 * `input[type=datetime-local]`. Совпадение «сегодня, но время уже прошло» уходит на следующую
 * неделю: дата встречи обязана быть в будущем (бэкенд отклонит прошедшую с 400).
 *
 * Возвращает пустую строку, если день недели не задан — поле даты остаётся пустым, как и
 * при создании с нуля.
 */
export function nextOccurrenceLocal(
  weekday: number | null,
  time: string | null,
  now: Date = new Date(),
): string {
  if (weekday === null) return '';

  const [hours = 0, minutes = 0] = (time ?? '00:00:00').split(':').map(Number);
  const target = new Date(now);
  target.setHours(hours, minutes, 0, 0);

  const daysAhead = (weekday - isoWeekdayOf(now) + 7) % 7;
  target.setDate(target.getDate() + daysAhead);
  // daysAhead > 0 уже даёт будущий день; догонять неделю нужно только когда совпал сегодняшний
  // день недели, а час встречи уже прошёл.
  if (target.getTime() <= now.getTime()) target.setDate(target.getDate() + 7);

  return toDatetimeLocalValue(target);
}

/**
 * Встреча → тело сохранения шаблона. Дата не переносится: вместо неё выводятся день недели и
 * время. Имя предзаполняется названием встречи — организатор чаще всего просто подтверждает.
 */
export function eventToTemplateBody(
  event: {
    title: string;
    description: string | null;
    locationText: string | null;
    locationLat: number | null;
    locationLon: number | null;
    locationHint: string | null;
    participantLimit: number | null;
    format: EventFormat;
    stage2LeadMinutesOverride: number | null;
    photoUrl: string | null;
    eventDatetime: string;
  },
  name: string,
): SaveEventTemplateBody {
  const startsAt = new Date(event.eventDatetime);
  return {
    name,
    title: event.title,
    description: event.description,
    locationText: event.locationText,
    locationLat: event.locationLat,
    locationLon: event.locationLon,
    locationHint: event.locationHint,
    participantLimit: event.participantLimit,
    format: event.format,
    // Берём СОБСТВЕННЫЙ интервал события, а не эффективный: подставленный сервером дефолт,
    // записанный в шаблон, молча стал бы явным выбором организатора и перестал бы следовать
    // за настройкой бэкенда (тот же урок, что у формы редактирования встречи).
    // У формата «сколько придёт» своего интервала не бывает — бэкенд отклонит.
    stage2LeadMinutes: event.format === 'any' ? null : event.stage2LeadMinutesOverride,
    photoUrl: event.photoUrl,
    defaultWeekday: isoWeekdayOf(startsAt),
    defaultTime: localTimeOf(startsAt),
  };
}

/**
 * Шаблон → тело сохранения. Нужен там, где шаблон перезаписывается целиком (переименование,
 * «Обновить шаблон»): PUT — полная замена, а серверные поля (`id`, `clubName`, даты) в теле
 * лишние. Явный перенос вместо спреда DTO: добавится поле — компилятор укажет сюда.
 */
export function templateToSaveBody(
  template: EventTemplateDto,
  overrides: Partial<SaveEventTemplateBody> = {},
): SaveEventTemplateBody {
  return {
    name: template.name,
    title: template.title,
    description: template.description,
    locationText: template.locationText,
    locationLat: template.locationLat,
    locationLon: template.locationLon,
    locationHint: template.locationHint,
    participantLimit: template.participantLimit,
    format: template.format,
    stage2LeadMinutes: template.stage2LeadMinutes,
    photoUrl: template.photoUrl,
    defaultWeekday: template.defaultWeekday,
    defaultTime: template.defaultTime,
    ...overrides,
  };
}

